File: 143-reorderList.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param {ListNode} head
    # @return {void} Do not return anything, modify head in-place instead.
    def reorderList(self, head):
        # find the middle of the linked list
        if head is None or head.next is None:
           return
        
        fast = slow = head
        while fast.next and fast.next.next:
              slow = slow.next
              fast = fast.next.next
        
        # seperate right and left linked list
        mid       = slow.next
        slow.next = None        
        left      = head
        right     = mid

        # reverse the right part of 
        start = None
        while right:
              next       = right.next
              right.next = start
              start      = right
              right      = next
        right = start
        
        # merge two linked list together
        new_start = new = ListNode(0)
        while left:
              new.next = left
              new = new.next
              left = left.next
              if right:
                 new.next = right
                 new = new.next
                 right = right.next
        head = new_start.next

    def print_llst(self, head):
        llst = ""
        while head:
              llst += str(head.val)
              if head.next:
                 llst += '->'
              head = head.next
        print 'llst: ' + llst

if __name__ == '__main__':
   s = Solution()
   test = ListNode(1)
   test.next = ListNode(2)
   test.next.next = ListNode(3)
   test.next.next.next = ListNode(4)
   test.next.next.next.next = ListNode(5)
   test.next.next.next.next.next = ListNode(6)
   test.next.next.next.next.next.next = ListNode(7)
   s.reorderList(test)
   s.print_llst(test)




        

File: C++/143-reorderList.cpp
#include "include/common.h"

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
         if (head == nullptr) {
            return;
         };
         ListNode *slow = head, *fast = head;
         ListNode *right, *left, *start, *nxt;
         // get the middle
         while(fast->next && fast->next->next) {
            fast = fast->next->next;
            slow = slow->next;
         };
         right      = slow->next;
         slow->next = nullptr;
         left       = head;
         //reverse right part
         start      = nullptr;
         while (right) {
            nxt         = right->next;
            right->next = start; // key steps!!
            start       = right;
            right       = nxt;
         };
         right = start; // key steps!!
         
         // merge two list together
         ListNode *new_head = new ListNode(0);
         ListNode *new_dummy = new_head;
         while (left) {
            new_dummy->next = left;
            left = left->next;
            new_dummy = new_dummy->next;
            if (right) {
                new_dummy->next = right;
                right = right->next;
                new_dummy = new_dummy->next;
            };
         };
         head = new_head->next;
    }
};

int main(void) {
    Solution s;
    Utils u;
    ListNode l1(1);
    ListNode l2(2);
    ListNode l3(3);
    ListNode l4(4);
    (&l1)->next = &l2;
    (&l2)->next = &l3;
    (&l3)->next = &l4;
    s.reorderList(&l1);
    u.printLinkedList(&l1);
    return 0;
};